### PR TITLE
[None][feat] AutoDeploy: improved sharding utilities

### PR DESCRIPTION
**Pull request by greg-kwasniewski1**
## Release Notes

* **Refactor**
  * Improved internal weight handling architecture to support multiple weights per operation, enabling more flexible model optimization
  * Enhanced quantization and sharding capabilities with improved parameter metadata tracking
  * Strengthened infrastructure for weight extraction and optimization callback mechanisms
  * Refactored parameter processing for consistent handling across quantization and sharding workflows

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Review by lucaslie — COMMENTED** [review]
Please add a tp>1 test here: https://github.com/NVIDIA/TensorRT-LLM/blob/74832a189589816fdafebc638aead5fc627989b7/tests/integration/defs/accuracy/test_llm_api_autodeploy.py#L221

Note that it must be correctly configured here: https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests/integration/test_lists/test-db

**Comment by greg-kwasniewski1** [comment]
@lucaslie regarding your comment:
> [Please add a tp>1 test here](https://github.com/NVIDIA/TensorRT-LLM/pull/10319#pullrequestreview-3618481804): 

I added it in [this separate PR](https://github.com/NVIDIA/TensorRT-LLM/pull/10364) that improves test coverage. Unless you want it in this PR?

File: tensorrt_llm/_torch/auto_deploy/models/custom/modeling_nemotron_h.py [hunk 1]
```diff
@@ -590,6 +582,13 @@ def __init__(self, config):
         self.backbone = NemotronHModel(config)
         self.vocab_size = config.vocab_size
         self.lm_head = nn.Linear(config.hidden_size, config.vocab_size, bias=False)
+        # Recursively iterate over all modules in self.backbone and list those with A_minus or A_log in their name
+        self.backbone_modules_with_A = []
+        for module_name, module in self.backbone.named_modules():
+            for param_name, _ in module.named_parameters(recurse=False):
+                if param_name in ("A_minus", "A_log"):
+                    self.register_load_state_dict_pre_hook(self._a_log_pre_hook)
+                    self.backbone_modules_with_A.append((module_name, param_name))
```
Comment (lucaslie) [inline tensorrt_llm/_torch/auto_deploy/models/custom/modeling_nemotron_h.py]: You only need to register this load state dict pre hook once since you already iterate over all the keys below:


```suggestion
        self.register_load_state_dict_pre_hook(self._a_log_pre_hook)
```

File: tensorrt_llm/_torch/auto_deploy/utils/node_utils.py [hunk 2]
```diff
@@ -129,8 +183,24 @@ def get_quantization_params_from_linear_node(linear_op: torch.fx.node.Node):
...
 
 
-def extract_weight_node(node: Node) -> int:
-    """Extracts the weight node from the given parametrized node"""
+def extract_weight_name(node: Node) -> str:
+    weight_nodes = extract_weight_nodes(node)
+    return weight_nodes.weights[0].node_key
+
+
+def get_const_tensor(tensor_name: str, gm: GraphModule) -> torch.Tensor:
```
Comment (lucaslie) [inline tensorrt_llm/_torch/auto_deploy/utils/node_utils.py]: misleading name? why are those "const tensors".

File: tests/integration/defs/accuracy/test_llm_api_autodeploy.py [hunk 3]
```diff
@@ -218,11 +218,13 @@ def test_bf16(self):
             task.evaluate(llm)
 
     @pytest.mark.skip_less_device_memory(32000)
-    def test_fp8(self):
+    @pytest.mark.parametrize("world_size", [1, 4])
```
Comment (lucaslie) [inline tests/integration/defs/accuracy/test_llm_api_autodeploy.py]: please correctly set up those tests in [`tests/integration/test_lists/test-db`](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests/integration/test_lists/test-db) Note that now that you parameterize over world_size you will have to call the correct test case for each of the single-gpu and multi-gpu test stages where you add this test

Comment (greg-kwasniewski1) [reply tests/integration/defs/accuracy/test_llm_api_autodeploy.py]: Removed this from this PR and moved it to[ this one.](https://github.com/NVIDIA/TensorRT-LLM/pull/10364)

File: tensorrt_llm/_torch/auto_deploy/utils/node_utils.py (hunk 2 above)
Comment (greg-kwasniewski1) [reply tensorrt_llm/_torch/auto_deploy/utils/node_utils.py]: Fixed

File: tensorrt_llm/_torch/auto_deploy/utils/node_utils.py [hunk 4]
```diff
@@ -325,6 +356,15 @@ def is_any_ssm_op(node: Node) -> bool:
...
 
+def is_any_conv_op(node: Node) -> bool:
+    return is_op(
+        node,
+        ops=[
+            torch.ops.auto_deploy.torch_causal_conv1d,
+        ],
+    )
+
+
```
Comment (galagam) [inline tensorrt_llm/_torch/auto_deploy/utils/node_utils.py]: Unused code?

**Review by galagam — APPROVED** [review]
LGTM

File: tensorrt_llm/_torch/auto_deploy/utils/node_utils.py [hunk 5]
```diff
@@ -155,57 +187,56 @@ def find_get_attr_node(weight_node: Node) -> Node:
...
+                    node_key=weight_node.target,
+                    tensor=get_param_or_buffer(weight_node.target, gm),
+                    submod=gm.get_submodule(weight_node.target.rpartition(".")[0]),
+                )
+            ],
+            biases=[],
+        )
     # for other parametrized nodes, we need to find the weight node
     else:
+        all_weight_nodes = [
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/auto_deploy/utils/node_utils.py]: Nit: could use the walrus operator to have a single call to `find_get_attr_node` per iteration:

```python
all_weight_nodes = [
    attr
    for n in node.args
    if isinstance(n, Node) and (attr := find_get_attr_node(n)) is not None
]
```

**Review by lucaslie — COMMENTED** [review]
please remove the A_log/A_minus updates since https://github.com/NVIDIA/TensorRT-LLM/pull/10512 will take care of this
